### PR TITLE
max zone size

### PR DESCRIPTION
**Pull request by nimaeskandary**
set a max for the number of records that can be in a zone to prevent possible out of memory errors

### CODE REVIEW

File: modules/api/src/main/resources/application.conf [hunk 1]
```diff
@@ -31,6 +31,8 @@ akka.http {
 vinyldns {
   data-stores = ["mysql", "dynamodb"]
 
+  max-zone-size = 100 # number of records that can be in a zone
```
Comment (nimaeskandary) [inline modules/api/src/main/resources/application.conf]: shouldnt be here

**Comment by nimaeskandary** [comment]
Screen Shot 2019-04-15 at 12 42 45 PM

**Comment by nimaeskandary** [comment]
Screen Shot 2019-04-15 at 12 42 31 PM

File: modules/api/src/it/resources/application.conf [hunk 2]
```diff
@@ -9,7 +9,7 @@ vinyldns {
   approved-name-servers = [
     "172.17.42.1."
   ]
-
+
```
Comment (mitruly) [inline modules/api/src/it/resources/application.conf]: Very subtle

File: modules/api/src/main/resources/reference.conf [hunk 3]
```diff
@@ -8,6 +8,8 @@ vinyldns {
   # if we should start up polling for change requests, set this to false for the inactive cluster
   processing-disabled = false
 
+  max-zone-size = 60000 # number of records that can be in a zone
```
Comment (mitruly) [inline modules/api/src/main/resources/reference.conf]: Wonder if `max-zone-sync-size` is clearer.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala [hunk 4]
```diff
@@ -70,10 +73,20 @@ case class DnsZoneViewLoader(zone: Zone, zoneTransfer: Zone => ZoneTransferIn)
           val rawDnsRecords: List[DNS.Record] =
             xfr.getAXFR.asScala.map(_.asInstanceOf[DNS.Record]).toList.distinct
 
-          val dnsZoneName = zoneDnsName(zone.name)
-          val recordSets = rawDnsRecords.map(toRecordSet(_, dnsZoneName, zone.id))
+          if (rawDnsRecords.length > maxZoneSize)
```
Comment (mitruly) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: From our initial discussion, we wanted to only include records that we are syncing in (excluding the dropped records); should probably block this until https://github.com/vinyldns/vinyldns/pull/582 is merged in (which brings the dropping to `ZoneViewLoader` from `ZoneSyncHandler`.

File: modules/api/src/it/resources/application.conf (hunk 2 above)
Comment (rebstar6) [reply modules/api/src/it/resources/application.conf]: i dig it

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala [hunk 5]
```diff
@@ -70,10 +73,20 @@ case class DnsZoneViewLoader(zone: Zone, zoneTransfer: Zone => ZoneTransferIn)
...
+            throw ZoneTooLargeError(
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: wow throw ew

Comment (rebstar6) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: also, we need to check on what the retry logic is here. this should *not* retry in the case that the zone is too large. we would want the sync to fail and a system message to indicate that in the zone change for this sync (read: sync message must be deleted from queue).

**Comment by mitruly** [comment]
Also was thinking that since we use a `diff` in `ZoneSyncHandler.scala` to identify if records currently in VinylDNS actually differ against the DNS backend, it might make more sense to apply the limit in `runSync` after we know how many actual record we're trying to write.

**Comment by nimaeskandary** [comment]
@mitruly Not so sure that would solve the issue, the zone view loader returns a list of records sets: https://github.com/vinyldns/vinyldns/blob/master/modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala#L74, so if we wait for https://github.com/vinyldns/vinyldns/blob/master/modules/api/src/main/scala/vinyldns/api/engine/ZoneSyncHandler.scala#L89 that may cause the out of mem error depending on the zone

**Comment by mitruly** [comment]
@nimaeskandary Another issue would be the following scenario:
- User syncs in zone for 60k records
- User makes a few changes
- Maybe some DNS records get created out-of-band (via DNS backend directly)
- User hits sync and hits an error where sync exceeds max limit

**Comment by mitruly** [comment]
With https://github.com/vinyldns/vinyldns/blob/ef30ce3fdc906c2b7ff0226d6d1dd548ed1f01d0/modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala#L76-L79 change, we shouldn't be deserializing all of those changes into memory, right?

**Comment by nimaeskandary** [comment]
At that point that we may be better off saying the zone is too big for vinyl by definition of our sync limit, because if the zone was abandoned they could not reconnect, because then the rules of whether a zone can be in or not would be more clear

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala (hunk 5 above)
Comment (nimaeskandary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: Yeah would have IO.raised but is already in an IO

Comment (nimaeskandary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: And here we catch throwables and just complete the zone change without syncing so the queue will not retry https://github.com/vinyldns/vinyldns/blob/master/modules/api/src/main/scala/vinyldns/api/engine/ZoneSyncHandler.scala#L152

File: modules/api/src/main/scala/vinyldns/api/VinylDNSConfig.scala [hunk 6]
```diff
@@ -112,4 +112,6 @@ object VinylDNSConfig {
     if (vinyldnsConfig.hasPath(key)) {
       vinyldnsConfig.getStringList(key).asScala.toList
     } else List()
+
+  lazy val maxZoneSize: Int = vinyldnsConfig.getInt("max-zone-size")
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/VinylDNSConfig.scala]: pull as an option and getOrElse 60000

File: modules/api/src/it/scala/vinyldns/api/domain/zone/ZoneViewLoaderIntegrationSpec.scala [hunk 7]
```diff
@@ -44,5 +44,14 @@ class ZoneViewLoaderIntegrationSpec extends WordSpec with Matchers {
           .load()
           .unsafeRunSync())
     }
+
+    "return a failure if the zone is larger than the max zone size" in {
+      assertThrows[ZoneTooLargeError](
+        DnsZoneViewLoader(Zone("vinyldns.", "test@test.com"))
+          .copy(maxZoneSize = 1)
```
Comment (rebstar6) [inline modules/api/src/it/scala/vinyldns/api/domain/zone/ZoneViewLoaderIntegrationSpec.scala]: i think this copy thing is just `DnsZoneViewLoader(Zone("vinyldns.", "test@test.com"), DnsZoneViewLoader.dnsZoneTransfer, 1)`

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala (hunk 5 above)
Comment (rebstar6) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: yuck. i still really dont like the throw, how crazy are things if we make that return an IO[Either] (or EitherT)?

File: modules/api/src/main/resources/reference.conf (hunk 3 above)
Comment (britneywright) [reply modules/api/src/main/resources/reference.conf]: should add this to the docs around configs, with more context on what it's intended for. It's just for checking the size of the zone during sync, not hard limiting the number of records that can exist in a zone.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala (hunk 5 above)
Comment (pauljamescleary) [reply modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: just wrap a few steps in `IO`, and then you can have one of the steps do the `IO.raiseError`, can use flatMap or a for comprehension...

```scala
for {
  rawDnsRecords <- IO(xxx)
  _ <- IO(ensureOr?)
  recordSets <- IO(yyy)
} yield ZoneView()
```

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala [hunk 8]
```diff
@@ -54,31 +56,38 @@ object DnsZoneViewLoader extends DnsConversions {
...
-            rawDnsRecords.filter(record => fromDnsRecordType(record.getType) != RecordType.UNKNOWN)
-
-          val dnsZoneName = zoneDnsName(zone.name)
-
-          val recordSets = supportedRecords.map(toRecordSet(_, dnsZoneName, zone.id))
-
-          ZoneView(zone, recordSets)
-        }
+        for {
+          rawDnsRecords <- IO {
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: IO split here feels arbitrary, and is hard to read. switch to:

```
  def load: () => IO[ZoneView] =
    () =>
      monitor("dns.loadZoneView") {
        for {
          zoneXfr <- IO {
            val xfr = zoneTransfer(zone)
            xfr.run()
            xfr.getAXFR.asScala.map(_.asInstanceOf[DNS.Record]).toList.distinct
          }
          rawDnsRecords = zoneXfr.filter(record => fromDnsRecordType(record.getType) != RecordType.UNKNOWN)
          _ <- if (rawDnsRecords.length > maxZoneSize)
            IO.raiseError(ZoneTooLargeError(zone, rawDnsRecords.length, maxZoneSize))
          else IO.pure(Unit)
          dnsZoneName <- IO(zoneDnsName(zone.name))
          recordSets = rawDnsRecords.map(toRecordSet(_, dnsZoneName, zone.id))
        } yield ZoneView(zone, recordSets)
    }
```

was not sure on last line (rawDnsRecords.map..). i dont think that needs to be wrapped in IO, but if so, wrap it.

for the error to work, need a :
```
object ZoneTooLargeError {
  def apply(zone: Zone, len: Int, maxZoneSize: Int): ZoneTooLargeError = {
    val msg = s"ZoneTooLargeError: Zone '${zone.name}' (id: '${zone.id}') contains $len records" +
      s"which exceeds the max of $maxZoneSize"
    new ZoneTooLargeError(msg)
  }
}
```

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala [hunk 9]
```diff
@@ -54,31 +56,31 @@ object DnsZoneViewLoader extends DnsConversions {
...
-          ZoneView(zone, recordSets)
-        }
+          }
+          rawDnsRecords = zoneXfr.filter(record =>
+            fromDnsRecordType(record.getType) != RecordType.UNKNOWN)
+          _ <- if (rawDnsRecords.length > maxZoneSize)
+            IO.raiseError(ZoneTooLargeError(zone, rawDnsRecords.length, maxZoneSize))
+          else IO.pure(Unit)
+          dnsZoneName <- IO(zoneDnsName(zone.name))
+          recordSets = rawDnsRecords.map(toRecordSet(_, dnsZoneName, zone.id))
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneViewLoader.scala]: sorry, actually, this can throw, so wrap in IO. otherwise 👍

**Pull request merged**